File: server/state.php
<?php
require_once 'player.php';
require_once 'card.php';

function next_player_key(array &$state): string {
    $currentIndex = array_search($state['acting_player'], $state['player_order']);
    if ($currentIndex === false) {
        return $state['player_order'][0];
    }
    return $state['player_order'][($currentIndex + 1) % count($state['player_order'])];
}

function remove_players(array &$state, array $player_keys) {
    $state['player_order'] = array_values(array_diff($state['player_order'], $player_keys));
}

function add_players(array &$state, array $player_keys) {
    $state['player_order'] = array_merge($player_keys, $state['player_order']);
}

function end_turn(array &$state) {
    $state['acting_player'] = next_player_key($state);
}

function add_energy(array &$state, int $add): bool {
    $player_key = acting_player($state);
    $state["$player_key.energy"] = $state["$player_key.energy"] ?? 0;
    $state["$player_key.energy"] += $add;
    if (pl_dotkey($state, 'energy') <= 0) {
        $state["$player_key.energy"] = 0;
        remove_players($state, ['anar']);
        return true;
    }
    $state["$player_key.energy"] = 
    pl_dotkey($state, 'energy') > $state["$player_key.max_energy"] ?
        $state["$player_key.max_energy"] :
        pl_dotkey($state, 'energy');
    
    return false;
}

function add_food(array &$state, int $add) {
    $loc = pl_dotkey($state, 'location');
    $key = "$loc.food";
    $state[$key] = ($state[$key] ?? 0) + $add;
    if ($state[$key] < 0) {
        $state[$key] = 0;
    }
}

function add_wood(array &$state, int $add) {
    $loc = pl_dotkey($state, 'location');
    $key = "$loc.wood";
    $state[$key] = ($state[$key] ?? 0) + $add;
    if ($state[$key] < 0) {
        $state[$key] = 0;
    }
}

function add_fire(array &$state, int $add) {
    $loc = pl_dotkey($state, 'location');
    $key = "$loc.fire_minutes";
    $state[$key] = ($state[$key] ?? 0) + $add;
    if ($state[$key] < 0) {
        $state[$key] = 0;
    }
}

function date_time_stamp(array $state) {
    return strtotime($state['date_time']);
}

function sun_light_level(array &$state) {
    $timestamp = date_time_stamp($state); 
    $lat = 51.4; $long = 0;
    $sun_info = date_sun_info($timestamp, $lat, $long);
    
    $sunrise = $sun_info['sunrise'];
    $sunset = $sun_info['sunset'];

    if ($timestamp > $sunrise && $timestamp < $sunset) {
        $day_length = $sunset - $sunrise;
        $progress = ($timestamp - $sunrise) / $day_length;
        $light_level = sin($progress * M_PI);

        return round($light_level * 10);
    } 

    return 0;
}

function light_level(array &$state) {
    $loc = pl_dotkey($state, 'location');
    $sun_light_level = sun_light_level($state);
    $fire_light_level = (($state["$loc.fire_minutes"] ?? 0) > 0) ? 4 : 0;
    return $sun_light_level + $fire_light_level;
}

function time_passes(int $minutes, array &$state) {
    $state['date_time'] = date('Y-m-d H:i:s', strtotime("+$minutes minutes", date_time_stamp($state)));
    
    foreach (human_keys() as $player_key) {
        $player_loc = $state["$player_key.location"];
        if (isset($state["$player_loc.fire_minutes"])) {
            $state["$player_loc.fire_minutes"] = max(0, $state["$player_loc.fire_minutes"] - $minutes);
        }
    }

    $loc = pl_dotkey($state, 'location');
    $energy_spent = round(-10 * ($minutes / 60));
    if ($state['sleeping'] ?? false) {
        $energy_spent = $energy_spent / 5;
    }
    if ($state["$loc.shelter"] ?? false) {
        $energy_spent = $energy_spent / 2;
    }

    return add_energy($state, $energy_spent);
}

function pl_dotkey(array $state, string $bare_key) {
    $player_key = acting_player($state);
    return $state["$player_key.$bare_key"];
}

function human_keys() {
    return ['anar', 'elshad'];
}

function init_state(): array {
    $state = [
        'status' => 'RUNNING',
        'round' => 1,
        'player_order' => array_values(array_merge(human_keys(), ['round'])),
        'acting_player' => human_keys()[0],

        'date_time' => date('Y-m-d 00:00:00'),
    ];

    foreach (human_keys() as $human_key) {
        $state["$human_key.location"] = 'forest';
        $state["$human_key.max_energy"] = 100;
        $state["$human_key.energy"] = 100;
    }

    return $state;
}

function robot_players(Socket $robot_socket) {
    return [
        'round' => new Player(
            'round',
            $robot_socket,
            function() {
                return robot_input('end_of_round');
            }
        )
    ];
}

function init_deck(): array {
    return [
        'skip' => new Card(
            'skip',
            function() { return true; },
            function(&$state) {
                end_turn($state);
                return "Let's see what else is happening.";
            }
        ),
        'eat' => new Card(
            'Eat 10 food',
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state["$loc.food"] ?? 0) > 0
                );
            },
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                $pdied = time_passes(30, $state);
                if($pdied) return "Player died from hunger.";
                $cooked = ($state["$loc.fire_minutes"] ?? 0) > 0 ? 50 : 20;
                $pdied = add_energy($state, $cooked);
                if($pdied) return "Player died from overeating.";
                add_food($state, -15);
                
                return $cooked > 20 ? "The player ate a cooked meal." : "The player ate raw food.";
            }
        ),
        'hunt' => new Card(
            'Hunt game',
            function(&$state) {
                return in_array(acting_player($state), human_keys());
            },
            function(&$state) {
                $light = sun_light_level($state);
                $is_day = ($light > 0);
                $yield = $is_day ? 12 : 3;
                add_food($state, $yield);
                $pdied = add_energy($state, -10);
                if($pdied) return "Player died from exhaustion.";
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";
                return $is_day ?
                "The daytime hunt was highly productive." :
                "Hunting at night was difficult.";
            }
        ),
        'wood' => new Card(
            'Collect wood',
            function(&$state) {
                return (
                    in_array(acting_player($state), human_keys()) &&
                    pl_dotkey($state, 'energy') > 5
                );
            },
            function(&$state) {
                $pdied = add_energy($state, -5);
                if($pdied) return "Player died from exhaustion.";
                $has_light = light_level($state) > 3;
                add_wood($state, ($has_light ? 5 : 3));
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";

                return $has_light ?
                "The player foraged for wood." :
                "The player foraged for wood. Lack of visibility made foraging challenging.";
            }
        ),
        'shelter' => new Card(
            'Build a shelter (50 wood)',
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state["$loc.wood"] ?? 0) >= 50
                );
            },
            function(&$state) {
                $loc = pl_dotkey($state, 'location');

                $pdied = add_energy($state, -45);
                if($pdied) return "Player died from exhaustion.";
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";

                $state["$loc.shelter"] = true;
                add_wood($state, -50);

                return "The player built a shelter.";
            }
        ),
        'boat' => new Card(
            'Build a boat (250 wood)',
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state["$loc.wood"] ?? 0) >= 250
                );
            },
            function(&$state) {
                $loc = pl_dotkey($state, 'location');

                $pdied = add_energy($state, -45);
                if($pdied) return "Player died from exhaustion.";
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";

                $state["$loc.boat"] = true;
                add_wood($state, -250);

                return "The player built a boat.";
            }
        ),
        'fish' => new Card(
            'Go fishing',
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state["$loc.boat"] ?? false)
                );
            },
            function(&$state) {
                $lucky = rand(0, 1);
                $yield = $lucky ? 12 : 3;
                add_food($state, $yield);
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";
                
                if (sun_light_level($state) > 6 && !isset($state['bottle_map'])) {
                    $state['bottle_map'] = true;
                    return "The player fishes out a map in a bottle!";
                }

                return $lucky ?
                "The player caught a big fish." :
                "The fishing trip was unlucky.";
            }
        ),
        'fire' => new Card(
            'Make fire (up to 10 wood)',
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state["$loc.wood"] ?? 0) > 0
                );
            },
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                $from_scratch = false;

                if (($state["$loc.fire_minutes"] ?? 0) <= 0) {
                    $from_scratch = true;
                    $pdied = add_energy($state, -45);
                    if($pdied) return "Player died from exhaustion.";
                    $pdied = time_passes(60, $state);
                    if($pdied) return "Player died from hunger.";
                }

                $current_wood = $state["$loc.wood"] ?? 0;
                $wood_to_burn = min($current_wood, 10);

                add_fire($state, round($wood_to_burn * 60 * 1.5));
                add_wood($state, -$wood_to_burn);

                return $from_scratch ? 
                "The player rubs sticks together to make fire. It was exhausting and time consuming." : 
                "The player stokes the fire with more wood.";
            }
        ),
        'follow_the_map' => new Card(
            'Follow the map the player fished out.',
            function (&$state) {
                return (
                    in_array(acting_player($state), human_keys()) &&
                    ($state['bottle_map'] ?? false)
                );
            },
            function (&$state) {
                $player_key = acting_player($state);
                $loc = pl_dotkey($state, 'location');
                $food = $state["$loc.food"];
                $wood = $state["$loc.wood"];
                $fire = $state["$loc.fire_minutes"];
                add_food($state, -10);
                add_wood($state, -10);
                add_fire($state, -60);

                $state["$player_key.location"] = 'caves';
                add_food($state, $food);
                add_wood($state, $wood);
                add_fire($state, $fire);

                return "The player enters the caves with all the supplies they could carry.";
            }
        ),
        'sleep' => new Card(
            'Sleep 8 hours',
            function(&$state) {
                return (
                    in_array(acting_player($state), human_keys()) &&
                    sun_light_level($state) <= 0
                );
            },
            function(&$state) {
                $loc = pl_dotkey($state, 'location');
                $state['sleeping'] = true;
                $has_fire = ($state["$loc.fire_minutes"] ?? 0) > 0;
                
                $pdied = time_passes(8 * 60, $state);
                if($pdied) return "Player died from hunger.";
                $pdied = add_energy($state, $has_fire ? 70 : 60);
                if($pdied) return "Player died from exhaustion.";
                unset($state['sleeping']);

                return $has_fire ?
                "The player slept in warmth." :
                "The player slept in the cold.";
            }
        ),
        'wait' => new Card(
            'Wait 1 hour',
            function(&$state) { return in_array(acting_player($state), human_keys()); },
            function(&$state) {
                $pdied = time_passes(60, $state);
                if($pdied) return "Player died from hunger.";
                return "Player does nothing for 1 hour.";
            }
        ),
        'end_of_round' => new Card(
            'End of the round',
            function(&$state) {
                return acting_player($state) === 'round';
            },
            function(&$state) {
                $state['round']++;
                end_turn($state);
                return "End of the round.";
            }
        ),
    ];
}

File: server-go/mods/vanilla/state.go
package vanilla

import (
	"game/server/mods/shared"
	"math"
	"math/rand"
	"net"
	"slices"
	"time"
)

type Player struct {
	Name   string
	Conn   net.Conn
	Decide func(state *shared.GameState) string
}

type Card struct {
	Name       string
	Conditions func(state *shared.GameState) bool
	Action     func(state *shared.GameState)
}

func getInt(state *shared.GameState, key string) int {
	if val, ok := state.Data[key].(int); ok {
		return val
	}
	return 0
}

func getString(state *shared.GameState, key string) string {
	if val, ok := state.Data[key].(string); ok {
		return val
	}
	return ""
}

func getStringList(state *shared.GameState, key string) []string {
	if val, ok := state.Data[key].([]string); ok {
		return val
	}
	return []string{}
}

func getBool(state *shared.GameState, key string) bool {
	if val, ok := state.Data[key].(bool); ok {
		return val
	}
	return false
}

func humanKeys() []string {
	return []string{"anar", "elshad"}
}

func robotKeys() []string {
	keys := []string{}
	for k := range InitRobots() {
		keys = append(keys, k)
	}
	return keys
}

func isHuman(p string) bool {
	return slices.Contains(humanKeys(), p)
}

func PropedSet(s *shared.GameState, key string, val int) {
	if val < 0 {
		val = 0
	}
	s.Data[key] = val
}

func GetActingPlayer(s *shared.GameState) string {
	return getString(s, "acting_player")
}

func PlayerDotKey(s *shared.GameState, key string) string {
	p := GetActingPlayer(s)
	return p + "." + key
}

func LocalDotKey(s *shared.GameState, key string) string {
	loc := getString(s, PlayerDotKey(s, "location"))
	return loc + "." + key
}

func LogEvent(s *shared.GameState, msg string) {
	logs := getStringList(s, "event_logs")
	s.Data["event_logs"] = append(logs, msg)
}

func EndTurn(s *shared.GameState) {
	order := getStringList(s, "player_order")
	current := GetActingPlayer(s)
	nextIdx := 0
	for i, name := range order {
		if name == current {
			nextIdx = (i + 1) % len(order)
			break
		}
	}
	s.Data["acting_player"] = order[nextIdx]
}

func RemovePlayers(s *shared.GameState, playerKeys []string) {
	order := getStringList(s, "player_order")

	var newOrder []string
	for _, player := range order {
		if !slices.Contains(playerKeys, player) {
			newOrder = append(newOrder, player)
		}
	}

	s.Data["player_order"] = newOrder
}

func AddEnergy(s *shared.GameState, amount int, deathMsg string) bool {
	key := PlayerDotKey(s, "energy")
	
	current := getInt(s, key)
	max := getInt(s, PlayerDotKey(s, "max_energy"))
	
	newVal := current + amount
	if newVal <= 0 {
		PropedSet(s, key, 0)
		LogEvent(s, deathMsg)
		RemovePlayers(s, []string{GetActingPlayer(s)})
		EndTurn(s)
		return true
	}
	
	if newVal > max {
		newVal = max
	}
	PropedSet(s, key, newVal)
	return false
}

func SunLightLevel(s *shared.GameState) int {
	t, _ := time.Parse("2006-01-02 15:04:05", getString(s, "date_time"))
	hour := t.Hour()
	
	// Simplified sun logic: Peak at 12:00, 0 at night
	if hour >= 6 && hour <= 18 {
		progress := float64(hour-6) / 12.0
		level := math.Sin(progress * math.Pi)
		return int(math.Round(level * 10))
	}
	return 0
}

func LightLevel(s *shared.GameState) int {
	sunLightLevel := SunLightLevel(s)

	fire_light_level := 0
	if getInt(s, LocalDotKey(s, "fire_minutes")) > 0 {
		fire_light_level = 4
	}

	return sunLightLevel + fire_light_level
}

func TimePasses(s *shared.GameState, minutes int) bool {
	t, _ := time.Parse("2006-01-02 15:04:05", getString(s, "date_time"))
	newTime := t.Add(time.Duration(minutes) * time.Minute)
	s.Data["date_time"] = newTime.Format("2006-01-02 15:04:05")

	// 2. Decay fires across all human player locations
	humans := []string{"anar", "elshad"}
	for _, pKey := range humans {
		playerLoc := getString(s, pKey+".location")
		if playerLoc != "" {
			fireKey := playerLoc + ".fire_minutes"
			currentFire := getInt(s, fireKey)
			if currentFire > 0 {
				newFire := currentFire - minutes
				if newFire < 0 {
					newFire = 0
				}
				PropedSet(s, fireKey, newFire)
			}
		}
	}
	baseEnergySpent := -10.0 * (float64(minutes) / 60.0)
	if getBool(s, PlayerDotKey(s, "sleeping")) {
		baseEnergySpent /= 5.0
	}

	if getBool(s, LocalDotKey(s, "shelter")) {
		baseEnergySpent /= 2.0
	}

	if getInt(s, LocalDotKey(s, "fire_minutes")) > 0 {
		baseEnergySpent /= 2.0
	}

	finalEnergySpent := int(math.Round(baseEnergySpent))
	return AddEnergy(s, finalEnergySpent, "Player died of hunger.")
}

func ClearEventLogs(s *shared.GameState) {
	s.Data["event_logs"] = []string{}
}

func InitState() *shared.GameState {
	return &shared.GameState{
		Data: map[string]interface{}{
			"status":		"RUNNING",
			"event_logs":	[]string{},
			"invisible_keys": []string{"invisible_keys", "event_logs"},
			"round":		 1,
			"player_order":  slices.Concat(humanKeys(), robotKeys()),
			"acting_player": "anar",
			"date_time":	 time.Now().Format("2006-01-02 00:00:00"),
			"anar.location": "forest",
			"anar.energy":   100,
			"anar.max_energy": 100,
			"elshad.location": "forest",
			"elshad.energy":   100,
			"elshad.max_energy": 100,
			"forest.wood": 0,
			"forest.food": 0,
		},
	}
}

func InitRobots() map[string]*Player {
	return map[string]*Player{
		"round": {
			Name: "round",
			Decide: func(s *shared.GameState) string {
				return RobotInput("end_of_round")
			},
		},
	}
}

func InitDeck(s *shared.GameState) map[string]*Card {
	return map[string]*Card{
		"skip": {
			Name: "Skip",
			Conditions: func(state *shared.GameState) bool { return true },
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				EndTurn(state)
				LogEvent(state, "Let's see what else is happening.")
			},
		},
		"eat": {
			Name: "Eat 10 food",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getInt(state, LocalDotKey(state, "food")) > 0
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)

				if TimePasses(state, 30) {
					return
				}

				cooked := 20
				if getInt(state, LocalDotKey(state, "fire_minutes")) > 0 {
					cooked = 50
				}

				if AddEnergy(state, cooked, "Died eating") {
					return
				}

				PropedSet(state, LocalDotKey(state, "food"), getInt(state, LocalDotKey(state, "food")) - 15)

				message := "The player ate raw food."
				if cooked > 20 {
					message = "The player ate a cooked meal."
				}
				LogEvent(state, message)
			},
		},
		"hunt": {
			Name: "Hunt game",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state))
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				isDay := SunLightLevel(state) > 0
				yield := 3
				if isDay {
					yield = 12
				}
				PropedSet(state, LocalDotKey(state, "food"), getInt(state, LocalDotKey(state, "food")) + yield)

				if AddEnergy(state, -10, "Died hunting") {
					return
				}
				if TimePasses(state, 60) {
					return
				}

				message := "Hunting at night was difficult."
				if isDay {
					message = "The daytime hunt was highly productive."
				}
				LogEvent(state, message)
			},
		},
		"wood": {
			Name: "Collect wood",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getInt(state, PlayerDotKey(state, "energy")) > 5
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				if AddEnergy(state, -5, "Died collecting wood") {
					return
				}

				hasLight := LightLevel(state) > 3
				yield := 3
				if hasLight {
					yield = 5
				}

				PropedSet(state, LocalDotKey(state, "wood"), getInt(state, LocalDotKey(state, "wood")) + yield)

				if TimePasses(state, 60) {
					return
				}

				message := "The player foraged for wood. Lack of visibility made foraging challenging."
				if hasLight {
					message = "The player foraged for wood."
				}
				LogEvent(state, message)
			},
		},
		"shelter": {
			Name: "Build a shelter (50 wood)",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getInt(state, LocalDotKey(state, "wood")) >= 50
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)

				if AddEnergy(state, -45, "Died building shelter") {
					return
				}
				if TimePasses(state, 60) {
					return
				}

				state.Data[LocalDotKey(state, "shelter")] = true
				PropedSet(state, LocalDotKey(state, "wood"), getInt(state, LocalDotKey(state, "wood")) - 50)

				LogEvent(state, "The player built a shelter.")
			},
		},
		"boat": {
			Name: "Build a boat (250 wood)",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getInt(state, LocalDotKey(state, "wood")) >= 250
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)

				if AddEnergy(state, -45, "Died building boat") {
					return
				}
				if TimePasses(state, 60) {
					return
				}

				state.Data[LocalDotKey(state, "boat")] = true
				PropedSet(state, LocalDotKey(state, "wood"), getInt(state, LocalDotKey(state, "wood")) - 250)

				LogEvent(state, "The player built a boat.")
			},
		},
		"fish": {
			Name: "Go fishing",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getBool(state, LocalDotKey(state, "boat"))
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				lucky := rand.Intn(2) == 1
				yield := 3
				if lucky {
					yield = 12
				}

				PropedSet(state, LocalDotKey(state, "food"), getInt(state, LocalDotKey(state, "food")) + yield)

				if TimePasses(state, 60) {
					return
				}

				if SunLightLevel(state) > 6 && !getBool(state, "bottle_map") {
					state.Data["bottle_map"] = true
					LogEvent(state, "The player fishes out a map in a bottle!")
					return
				}

				message := "The fishing trip was unlucky."
				if lucky {
					message = "The player caught a big fish."
				}
				LogEvent(state, message)
			},
		},
		"fire": {
			Name: "Make fire (up to 10 wood)",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getInt(state, LocalDotKey(state, "wood")) > 0
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				fromScratch := false

				if getInt(state, LocalDotKey(state, "fire_minutes")) <= 0 {
					fromScratch = true
					if AddEnergy(state, -45, "Died making fire from scratch") {
						return
					}
					if TimePasses(state, 60) {
						return
					}
				}

				currentWood := getInt(state, LocalDotKey(state, "wood"))
				woodToBurn := currentWood
				if woodToBurn > 10 {
					woodToBurn = 10
				}

				burnMinutes := woodToBurn * 90
				PropedSet(state, LocalDotKey(state, "fire_minutes"), getInt(state, LocalDotKey(state, "fire_minutes")) + burnMinutes)
				PropedSet(state, LocalDotKey(state, "wood"), currentWood - woodToBurn)

				message := "The player stokes the fire with more wood."
				if fromScratch {
					message = "The player rubs sticks together to make fire. It was exhausting and time consuming."
				}
				LogEvent(state, message)
			},
		},
		"follow_the_map": {
			Name: "Follow the map the player fished out.",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && getBool(state, "bottle_map")
			},
			Action: func(state *shared.GameState) {

				food := getInt(state, LocalDotKey(state, "food"))
				wood := getInt(state, LocalDotKey(state, "wood"))
				fire := getInt(state, LocalDotKey(state, "fire_minutes"))

				PropedSet(state, LocalDotKey(state, "food"), food - 10)
				PropedSet(state, LocalDotKey(state, "wood"), wood - 10)
				PropedSet(state, LocalDotKey(state, "fire_minutes"), fire - 60)

				newLoc := "caves"
				state.Data[PlayerDotKey(state, "location")] = newLoc

				PropedSet(state, newLoc+".food", getInt(state, newLoc+".food") + food)
				PropedSet(state, newLoc+".wood", getInt(state, newLoc+".wood") + wood)
				PropedSet(state, newLoc+".fire_minutes", getInt(state, newLoc+".fire_minutes") + fire)

				LogEvent(state, "The player enters the caves with all the supplies they could carry.")
			},
		},
		"sleep": {
			Name: "Sleep 8 hours",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state)) && SunLightLevel(state) <= 0
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)

				state.Data[PlayerDotKey(state, "sleeping")] = true
				hasFire := getInt(state, LocalDotKey(state, "fire_minutes")) > 0

				if TimePasses(state, 8 * 60) {
					return
				}

				delete(state.Data, PlayerDotKey(state, "sleeping"))

				message := "The player slept in the cold."
				if hasFire {
					message = "The player slept in warmth."
				}
				LogEvent(state, message)
			},
		},
		"wait": {
			Name: "Wait 1 hour",
			Conditions: func(state *shared.GameState) bool {
				return isHuman(GetActingPlayer(state))
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				if TimePasses(state, 60) {
					return
				}
				LogEvent(state, "Player does nothing for 1 hour.")
			},
		},
		"end_of_round": {
			Name: "End of the round",
			Conditions: func(state *shared.GameState) bool {
				return GetActingPlayer(state) == "round"
			},
			Action: func(state *shared.GameState) {
				ClearEventLogs(state)
				PropedSet(state, "round", getInt(state, "round") + 1)
				EndTurn(state)
				LogEvent(state, "End of the round.")
			},
		},
	}
}